### PR TITLE
Add aliasing support for custom predictable Lambda URLs

### PR DESCRIPTION
**Pull request by gregfurman**
## Motivation
- This PR is a followup to https://github.com/localstack/localstack/pull/11103 that adds support for including an alias when constructing a custom predictble URL

## Changes
- Including an existing `alias` in the `Qualifier` parameter when running a [CreateFunctionUrlConfig](https://docs.aws.amazon.com/lambda/latest/api/API_CreateFunctionUrlConfig.html) will create a compound custom URL in the form `<custom-id-value>-<qualifier-or-alias>`.
- Remove comment from previous PR outlining noting to add this functionality.

### Testing
- Ran all `only_localstack` tests in the `TestLambdaUrl` suite.
- Tested against local instance using `awslocal`.

## TODO

What's left to do:

- [x] Update the [Lambda docs](https://docs.localstack.cloud/user-guide/aws/lambda/) to include these changes in a follow-up PR.
- [x] Currently, deleting an alias does not remove it's corresponding URL (verified via CLI) -- there should be a parity test checking whether this is intended functionality.

### CODE REVIEW

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 1]
```diff
@@ -4102,6 +4103,72 @@ def test_create_url_config_custom_id_tag_invalid_id(
...
+            assert f"://{expected_url_id}.lambda-url." in url_config_created["FunctionUrl"]
+
+        def _assert_create_aliased_function_url(fn_version: str, fn_alias: str):
+            aws_client.lambda_.create_alias(
+                FunctionName=function_name, FunctionVersion=fn_version, Name=fn_alias
+            )
+            _assert_create_function_url(fn_alias, f"{custom_id_value}-{fn_alias}")
+
+        # Publishes a new version and creates an aliased URL
+        version = aws_client.lambda_.update_function_code(
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:4138]: That's an async operation and requires a waiter such as:
```
aws_client.lambda_.get_waiter("function_updated_v2").wait(FunctionName=function_name)
```

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 2]
```diff
@@ -4102,6 +4103,72 @@ def test_create_url_config_custom_id_tag_invalid_id(
...
+            function_url = aws_client.lambda_.get_function_url_config(
+                FunctionName=function_name, Qualifier=alias
+            ).get("FunctionUrl")
+            assert f"://{custom_id_value}-{alias}.lambda-url." in function_url
+
+        # Finally, check if the none-aliased URL can be retrieved
+        function_url = aws_client.lambda_.get_function_url_config(FunctionName=function_name).get(
+            "FunctionUrl"
+        )
+        assert f"://{custom_id_value}.lambda-url." in function_url
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply tests/aws/services/lambda_/test_lambda_api.py:4170]: I was actually also getting issues with invoking functions on my local. Assumed it had to do with my local setup.  Let me try again and update this thread.

[the remark this answers is not in the document]
Comment (gregfurman) [reply tests/aws/services/lambda_/test_lambda_api.py:4170]: Ahh, I think I found the issue. I wasn't adding a `Qualifier` to the `add_permissions` request. Since this is an aliased URL, the permission needs to be more fine-grained:

```
     aws_client.lambda_.add_permission(
            FunctionName=function_name,
            StatementId="urlPermission",
            Action="lambda:InvokeFunctionUrl",
            Principal="*",
            FunctionUrlAuthType="NONE",
            Qualifier=alias_name, # <-- this kwarg
        )
```

[the remark this answers is not in the document]
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_api.py:4170]: We should add a behavior test case to cover that scenario. Basically, testing the alias-based custom function URL analogous to the base custom_id feature here: `aws.services.lambda_.test_lambda.TestLambdaURL.test_lambda_url_invocation_custom_id`

[the remark this answers is not in the document]
Comment (gregfurman) [reply tests/aws/services/lambda_/test_lambda_api.py:4195]: Added a `test_lambda_url_invocation_custom_id_aliased` test for this 👍

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 3]
```diff
@@ -4102,6 +4103,72 @@ def test_create_url_config_custom_id_tag_invalid_id(
...
+        version = aws_client.lambda_.update_function_code(
+            FunctionName=function_name, ZipFile=zip_contents, Publish=True
+        ).get("Version")
+        _assert_create_aliased_function_url(fn_version=version, fn_alias="v1")
+
+        # Alias the $LATEST version
+        _assert_create_aliased_function_url(fn_version="$LATEST", fn_alias="latest")
+
+        # Update the code, creating an unpublished version
+        aws_client.lambda_.update_function_code(FunctionName=function_name, ZipFile=zip_contents)
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply tests/aws/services/lambda_/test_lambda_api.py]: Good idea. I added a check to ensure the contents of the lambda functions are identical and the revision ID differs between responses:

https://github.com/localstack/localstack/blob/b2c27b181ad09af05aded6aaf50c3f4c84549eb7/tests/aws/services/lambda_/test_lambda_api.py#L4165-L4169

https://github.com/localstack/localstack/blob/b2c27b181ad09af05aded6aaf50c3f4c84549eb7/tests/aws/services/lambda_/test_lambda_api.py#L4171-L4174

[the remark this answers is not in the document]
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_api.py]: ah, now I see. I missed the waiter within the assert helper. That should be sufficient.

**Review by joe4dev — APPROVED** [review]
Thorough test coverage covering multiple versions, logical follow-up exceptions, and even potential regressions on AWS features ✨
